### PR TITLE
Pull the common parts of NDArrays into the base NDArray class.

### PR DESCRIPTION
**Pull request by thetorpedodog**
Eliminates some duplicate code and some redundant type-assertions/casts.

### CODE REVIEW

**Comment by mlin** [comment]
@thetorpedodog Is there supposed to be a bigger diff/deletion to `sparse_nd_array.py` like there is in `dense_nd_array.py`?

**Comment by thetorpedodog** [comment]
> @thetorpedodog Is there supposed to be a bigger diff/deletion to `sparse_nd_array.py` like there is in `dense_nd_array.py`?

That part already happened in https://github.com/single-cell-data/TileDB-SOMA/commit/93050f1ca76700b6038033befe953b6f6085bee0#diff-ff1513d37e7cf18c204fe02ac36f85121c928794afc8012e438fe3cba63b2a2e. Unfortunately, there’s nothing more in common left to extract.

File: apis/python/src/tiledbsoma/sparse_nd_array.py [hunk 1]
```diff
@@ -26,7 +25,7 @@
 _UNBATCHED = options.BatchSize()
 
 
-class SparseNDArray(TileDBArray, somacore.SparseNDArray):
+class SparseNDArray(NDArray, somacore.SparseNDArray):
```
Comment (atolopko-czi) [inline apis/python/src/tiledbsoma/sparse_nd_array.py:28]: Can the `create()` class method be removed from `SparseNDArray` as well?

File: apis/python/src/tiledbsoma/common_nd_array.py [hunk 2]
```diff
@@ -3,15 +3,54 @@
...
 
 from . import util, util_arrow
 from .options.soma_tiledb_context import SOMATileDBContext
 from .options.tiledb_create_options import TileDBCreateOptions
+from .tiledb_array import TileDBArray
+
+_Self = TypeVar("_Self", bound="NDArray")
+
+
+class NDArray(TileDBArray, somacore.NDArray):
```
Comment (atolopko-czi) [inline apis/python/src/tiledbsoma/common_nd_array.py:22]: could this be documented as being a base class and is useful for polymorphic typing?

File: apis/python/src/tiledbsoma/sparse_nd_array.py (hunk 1 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/sparse_nd_array.py:26]: How did I manage to miss that? And the other things that are also shared in common? Feeling pretty silly about that right now.

File: apis/python/src/tiledbsoma/common_nd_array.py (hunk 2 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/common_nd_array.py:19]: Added a brief note. If you think it needs a little more context, please do say so.

**Review by atolopko-czi — APPROVED** [review]
LGTM!

**Pull request merged**